pahfit/fitter.py

```diff
@@ -0,0 +1,209 @@
+    dictionaries could also be used if absolutely necessary.
+
+    The main functionalities of a Fitter subclass:
+    1. Convert the numbers that are in the Features table to a fittable
+       model configuration for a certain framework. The details of the
+       fitting framework are hidden behind the respective subclass.
+    2. Fit the model to the spectrum without any additional assumptions.
+       The Fitter will fit the given data using the given model without
+       thinking about redshift, units, instrumental effects.
+    3. Retrieve the fitted quantities, which are the values that were
+       passed during step 1. When fit result uncertainties are
+       implemented, they will also need to be retrieved through this
+       API.
+    4. Access to the evaluation of the underlying model (again with no
+       assumptions like in step 2.).
+
+    A few notes on how the above is achieved:
+
+    For the model setup, there is one function per type of component
+    supported by PAHFIT, and the arguments of these functions will ask
+    for certain standard PAHFIT quantities (in practice, these are the
+    values stored in the Features table). The abstract Fitter class
+    ensure that the function signatures are the same between different
+    Fitter implementations, so that only a single logic has to be
+    implemented to up the Fitter (in practice, this is a loop over the
+    Features table implemented in Model).
+
+    During the Fitter setup, the initial values, bounds, and "fixed"
+    flags are passed using one function call for each component, e.g.
+    register_line(). Once all components have been added, the
+    finalize_model() function should be called; some subclasses (e.g.
+    APFitter) need to consolidate the registered components to prepare
+    the model that they manage for fitting. After this, fit() can be
+    called to apply the model and the astropy fitter to the data. The
+    results will then be retrievable for one component at a time, by
+    passing the component name to get_result().
+
+    """
+
+    @abstractmethod
+    def clear(self):
+        """Reset model.
+
+        After reset, register_() and finalize_model() can be used again.
+
+        """
+        pass
+
+    @abstractmethod
+    def components(self):
+        """Return list of features.
+
+        Only works after finalize_model(). Will return the names passed
+        using the register functions.
+
+        """
+        pass
+
+    @abstractmethod
+    def finalize_model(self):
+        """Process the registered features and prepare for fitting.
+
+        The register functions below allow adding individual features.
+        The exact implementation of how features are added, and
+        finalized in to a single fittable model, depend on the
+        underlying implementation.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_starlight(self, name, temperature, tau):
```

Comment:
OK this is cheeky, but `register_` really doesn't do it for me.  How about `add_feature_*`?

Reply:
Sounds good to me.

In the `APFitter` implementation, I'm also changing the wording. I use 'component' when using astropy lingo, i.e. everything related to setting up the multi component model (so only for hidden functions, variable names, and implementation comments).  But "feature" for the PAHFIT concepts.